### PR TITLE
Display BigFloat's based on their precision

### LINKED ISSUE #11565: set_bigfloat_precision behaves odd (or BigFloat display)
Something odd may be happening with `set_bigfloat_precision`, or possibly with the way `BigFloat`s are displayed.

Consider the following:
```julia
julia> VERSION # two days old...
v"0.4.0-dev+5147"

julia> get_bigfloat_precision()
256

julia> parse(BigFloat, "1.1")  # default behavior, which I think is ok
1.100000000000000000000000000000000000000000000000000000000000000000000000000003

julia> set_bigfloat_precision(40)
40

julia> parse(BigFloat, "0.1") # ??!?
1.100000000000363797880709171295166015625000000000000000000000000000000000000000
```
In the last line I was expecting something like 1.1000000000004e+00, since it mimics the last example (of the arbitrary precision section) of the [documentation](http://docs.julialang.org/en/latest/manual/integers-and-floating-point-numbers/#arbitrary-precision-arithmetic). But I can't actually reproduce it:
```julia
julia> with_bigfloat_precision(40) do
       parse(BigFloat,"1.1")
       end
1.100000000000363797880709171295166015625000000000000000000000000000000000000000
```

### PR DESCRIPTION
**Pull request by p-zubieta**
Fixes #11565

### CODE REVIEW

File: base/mpfr.jl [hunk 1]
```diff
@@ -765,12 +765,15 @@ function with_bigfloat_precision(f::Function, precision::Integer)
 end
 
 function string(x::BigFloat)
-    lng::Int32 = 127
+    # In general, the number of decimal places needed to read back the number exactly is
+    # ceil(log(10, 2^precision(x))) or ceil(precision(x) * log10(2))
+    dp = ceil(Int32, precision(x) * 0.3010299956639812)
```
Comment (vtjnash) [inline base/mpfr.jl]: does this need to be `log10(big(2))`, or does the Float64 approximation work for all reasonable values of BigFloat that could be encountered here

Comment (pabloferz) [inline base/mpfr.jl]: I don't know what approximation is done for Float64.

**Comment by vtjnash** [comment]
the change looks good to me.

can you add a few tests to show that this is working (and so it can't break in the future without someone noticing).

File: base/mpfr.jl [hunk 2]
```diff
@@ -765,12 +765,15 @@ function with_bigfloat_precision(f::Function, precision::Integer)
 end
 
 function string(x::BigFloat)
-    lng::Int32 = 127
+    # In general, the number of decimal places needed to read back the number exactly is
+    # ceil(log(10, 2^precision(x))) or ceil(precision(x) * log10(2))
```
Comment (simonbyrne) [inline base/mpfr.jl]: Doesn't this need to be 1 higher? see, for example http://www.exploringbinary.com/number-of-digits-required-for-round-trip-conversions/ and http://www.exploringbinary.com/the-inequality-that-governs-round-trip-conversions-a-partial-proof/

Comment (pabloferz) [inline base/mpfr.jl]: I guess I should change the comment which is misleading. You're right in that the number should be 1 higher, but here we only need `ceil(precision(x) * log10(2))`. We call MPFR functions so the number is returned in scientific notation, so I just wanted the most sigfinicand digits except for the first in order to know how decimal places to display after the dot.

I'll change the comment.

File: base/mpfr.jl (hunk 1 above)
Comment (simonbyrne) [inline base/mpfr.jl]: I agree, this should be fine (in fact `0.3010299956639812` is slightly larger than `log10(big(2.0))`)

File: base/mpfr.jl (hunk 2 above)
Comment (simonbyrne) [inline base/mpfr.jl]: Ah, that makes sense.

**Comment by tkelman** [comment]
Could you squash the commits?

**Comment by pabloferz** [comment]
The failures seem unrelated.

**Comment by pabloferz** [comment]
Is this in good shape now?

**Pull request merged**